### ISSUE TITLE
Custom mergeFunction for ConstraintCollectors.toMap is not used

### ISSUE THREAD

**Issue by storm-TOS** [TimefoldAI/timefold-solver#994]
## Description
In both a realistic application and a minimal setup, when using a groupBy node with a toMap collector based on a custom merge function, the merge function is not called when I expect it to. Instead of merging entries with equal keys, only one of the entries is kept.

Furthermore, if some Move causes this entry to be retracted, it is possible for the map to become empty, which is completely paradoxical when grouping stream elements.

No exceptions are thrown whatsoever.

## To reproduce
Evaluate the following constraint for a planning solution with at least two entities.
```
public Constraint constraint(ConstraintFactory factory) {
    return factory.forEach(Entity.class)
            .groupBy(entity -> 0,
                    ConstraintCollectors.toMap(
                            entity -> 0,
                            entity -> entity,
                            (entity1, entity2) -> {
                                throw new RuntimeException();
                            }
                    )
            )
            .penalize(HardSoftScore.ONE_HARD)
            .asConstraint("constraint");
}
```
Since all entities end up in the same group with the same key, I expect them to be merged. However, no instance of RuntimeException is thrown, which must mean that the merge function is never called.

When explaining a solution with two entities that print as "A" and "B" respectively, the explanation reads:
```
Explanation of score (-1hard/0soft):
    Constraint matches:
        -1hard: constraint (constraint) has 1 matches:
            -1hard: justified with ([0, {0=B}])
    Indictments:
        -1hard: indicted with (0) has 1 matches:
            -1hard: constraint (constraint)
        -1hard: indicted with ({0=B}) has 1 matches:
            -1hard: constraint (constraint)
```
It seems that an expected entry `0=A` was simply overwritten with `0=B`.

## Further details
I used **Timefold Community 1.12**, but ran into the same issue in version 1.10.

The specific overload of toMap used above is the one with signature
```
public static <A, Key, Value> ai.timefold.solver.core.api.score.stream.uni.UniConstraintCollector<A, ?, java.util.Map<Key, Value>> toMap(
        java.util.function.Function<? super A, ? extends Key> keyMapper,
        java.util.function.Function<? super A, ? extends Value> valueMapper,
        java.util.function.BinaryOperator<Value> mergeFunction
)
```

However, the same issue applies to the overload with signature
```
public static <A, B, Key, Value> ai.timefold.solver.core.api.score.stream.bi.BiConstraintCollector<A, B, ?, java.util.Map<Key, Value>> toMap(
        java.util.function.BiFunction<? super A, ? super B, ? extends Key> keyMapper,
        java.util.function.BiFunction<? super A, ? super B, ? extends Value> valueMapper,
        java.util.function.BinaryOperator<Value> mergeFunction
)
```

## A small warning
Funnily enough, a fix for this bug (if I am correct in regarding it as such) would most likely be a **breaking change** for our current code, and in theory this could apply to other users as well. I actually wrote a merge function that only throws an exception, because I incorrectly assumed that duplicate keys can never occur. I only now realize that this assumption does not always hold in early stages of solving.

Therefore, if anything gets changed about this, I recommend documenting it well in the release notes.

**Comment by triceo** [comment]
Thanks for reporting, @storm-TOS - we'll take a look when time permits.

**Comment by triceo** [comment]
@storm-TOS I wasn't able to reproduce the described behavior. There are [tests](https://github.com/TimefoldAI/timefold-solver/blob/main/core/src/test/java/ai/timefold/solver/core/impl/score/stream/collector/uni/InnerUniConstraintCollectorsTest.java#L743) in our codebase which check that the merge function is being used, and if I replace the merge function with a function that throws an exception, the exception is thrown.

I'm going to need a minimal reproducer from you.

**Comment by storm-TOS** [comment]
Hi @triceo, very interesting. This should do the trick, I think.
[toMapBug.zip](https://github.com/user-attachments/files/16508110/toMapBug.zip)

**Comment by triceo** [comment]
Thanks for the reproducer!
THe fix will be released next week in Timefold Solver 1.13.0.

**Comment by storm-TOS** [comment]
You're welcome!
